Harmony/Public/Patch.cs

```diff
@@ -295,8 +321,8 @@ public MethodInfo PatchMethod
 			set
 			{
 				patchMethod = value;
-				methodToken = patchMethod.MetadataToken;
-				moduleGUID = patchMethod.Module.ModuleVersionId.ToString();
```

Comment:
Can you elaborate on these changes?

Reply:
Yes!  So I can't serialize "PatchMethod", so I instead pass a `null` value in the constructor.

https://github.com/nwestfall/Harmony/blob/e30e0a1452b8cd5ce8e1e07963a0b0294416b869/Harmony/Public/Patch.cs#L472

You get `NullReferenceExceptions` then when it tries to update `methodToken` and `moduleGUID`.

If you don't want me to edit the variable, I can just `null` check in the constructor.

Reply:
I understand the implications and the effect but not the cause. Why wouldn’t you be able to serialize a patch method?

Reply:
This is because they contain `System.Type` fields, which cannot be serialized.

If you do, you get the following exception
```
Unhandled exception. System.NotSupportedException: Serialization and deserialization of 'System.Type' instances are not supported and should be avoided since they can lead to security issues. Path: $.DeclaringType.
```

Example code: https://dotnetfiddle.net/xszwaE

Reply:
I did check and the `PatchMethod` gets "rebuilt" after the next `get` since there is a `null` check above it.  It only gets rebuilt though, if I set the `methodToken` and `moduleGUID`, which I can serialize.

Reply:
The original code never serialized the MethodInfo because in some .NET implementations you cannot serialize MethodBase or Type. Instead it only serializes the two aggregate fields (see 274 for the do-not-serialize annotation and 276+277 for the substitution fields.

I think you need to make your serialization do the same as it did before.

Reply:
So I don't serialize `patchMethod` or `PatchMethod` either.  I serialize `methodToken` (276) and `moduleGUID` (277).  I accomplish this by writing my own serializer for `Patch` and do it in the following lines.

https://github.com/pardeike/Harmony/blob/e30e0a1452b8cd5ce8e1e07963a0b0294416b869/Harmony/Public/Patch.cs#L486-L487

When I then read the serialization, I pass them into a new constructor.
https://github.com/pardeike/Harmony/blob/e30e0a1452b8cd5ce8e1e07963a0b0294416b869/Harmony/Public/Patch.cs#L472

`PatchMethod`/`patchMethod` is never serialized and I rely on the same two fields you do `276` and `277`.

I think the confusion is I'm setting `PatchMethod` as `null` cause it's in the constructor.  If you want to leave it as it, I can make a new constructor where I don't need to pass `PatchMethod` so that way I don't need to `null` check.

Reply:
This is the constructor I would update to not have `patch` and then not call the other constructor, but rather copy the logic.

https://github.com/pardeike/Harmony/blob/e30e0a1452b8cd5ce8e1e07963a0b0294416b869/Harmony/Public/Patch.cs#L369-L374

Would turn into

```c#
internal Patch(int index, string owner, int priority, string[] before, string[] after, bool debug, int methodToken, string moduleGUID)
{
        this.index = index;
	this.owner = owner;
	this.priority = priority == -1 ? Priority.Normal : priority;
	this.before = before ?? new string[0];
	this.after = after ?? new string[0];
	this.debug = debug;
	this.methodToken = methodToken;
	this.moduleGUID = moduleGUID;
}
```